### PR TITLE
Update to nixos-23.11

### PR DESCRIPTION
**Pull request by thufschmitt**
# Motivation

Update to the latest stable NixOS release.

I didn't try to deeply check anything, just updated the flake input (and removed a redundant fetch) and saw that the build and tests didn't break.

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

**Comment by edolstra** [comment]
This is failing because the `bear` package is marked as broken. (Don't know what that is, but we're using it in the clang devshell.)

**Comment by cole-h** [comment]
It's used for LSP (if you run your `make` commands with `bear`, it'll generate `compile_commands.json` that LSP servers use to find e.g. headers and whatnot).

It was marked broken in https://github.com/NixOS/nixpkgs/pull/250435#issuecomment-1706856418 because it fails to build on x86_64-darwin (it built fine on aarch64-darwin for me).

**Comment by Ericson2314** [comment]
I guess let's adjust that `meta.broken` to be more fine-grained? And then the dep here can be corresponding adjusted.

**Comment by SuperSandro2000** [comment]
That was already started in https://github.com/NixOS/nixpkgs/pull/279035

**Comment by lf-** [comment]
@thufschmitt can you have another go at this? bear is fixed now, and I now have a PR blocked on the C++ compiler being antique.

**Comment by thufschmitt** [comment]
> @thufschmitt can you have another go at this? bear is fixed now, and I now have a PR blocked on the C++ compiler being antique.

The fix hasn't been backported to 23.11 (I've just opened https://github.com/NixOS/nixpkgs/pull/291814 for that).
But that shouldn't block us, we can just disable `bear` on Darwin. Trying to build the thing right now

**Comment by thufschmitt** [comment]
@Ericson2314 the evaluation failures on Darwin (https://github.com/NixOS/nix/actions/runs/8064971474/job/22029891317?pr=9546) are above my pay grade, do you think you can have a look?

**Comment by thufschmitt** [comment]
OK, so I managed to fix the evaluation errors on Darwin (by just disabling everything that failed).

I'm quite bewildered by the evaluation performance between 23.05 and 23.11 though: evaluating the flake on 23.11 takes nearly 3x longer than on 23.05 (despite having removed some outputs)

```console
$ hyperfine --warmup 1 \
    "nix flake show nix/nixos-23.11 --option eval-cache false --all-systems" \                                                            
    "nix flake show nix --all-systems --option eval-cache false"
Benchmark 1: nix flake show nix/nixos-23.11 --option eval-cache false --all-systems
  Time (mean ± σ):     27.201 s ±  0.065 s    [User: 19.490 s, System: 2.875 s]
  Range (min … max):   27.100 s … 27.310 s    10 runs
 
Benchmark 2: nix flake show nix --all-systems --option eval-cache false
  Time (mean ± σ):      9.398 s ±  0.216 s    [User: 7.261 s, System: 1.046 s]
  Range (min … max):    9.054 s …  9.843 s    10 runs
 
Summary
  nix flake show nix --all-systems --option eval-cache false ran
    2.89 ± 0.07 times faster than nix flake show nix/nixos-23.11 --option eval-cache false --all-systems
```

any one knows what that could be due to?

**Comment by thufschmitt** [comment]
@infinisil there's a failure in the fileset tests on darwin: https://github.com/NixOS/nix/actions/runs/8069576505/job/22044909419?pr=9546 does that match anything already known?

**Comment by infinisil** [comment]
Turns out `\s` is not POSIX.. See https://github.com/NixOS/nixpkgs/pull/291933 for a fix

File: flake.nix [hunk 1]
```diff
@@ -404,8 +396,11 @@
             # Make bash completion work.
             XDG_DATA_DIRS+=:$out/share
           '';
+
           nativeBuildInputs = attrs.nativeBuildInputs or []
-            ++ lib.optional stdenv.cc.isClang pkgs.buildPackages.bear
+            # TODO: Remove the darwin check once
+            # https://github.com/NixOS/nixpkgs/pull/291814 is available
+            ++ lib.optional (stdenv.cc.isClang && !stdenv.buildPlatform.isDarwin) pkgs.buildPackages.bear
```
Comment (roberth) [inline flake.nix:403]: Indeed #9900 did not need it. :+1:

**Review by roberth — APPROVED** [review]

**Comment by infinisil** [comment]
Docker push is failing: https://github.com/NixOS/nix/actions/runs/8084099897/job/22090423845

This is caused by using Nix 2.13.3:
https://github.com/NixOS/nix/blob/587c7dcb2bc4e504159f527c7e776ede231bd0db/.github/workflows/ci.yml#L119

Which doesn't have `builtins.readFileType`, causing `lib.fileset` to not work correctly in pure evaluation mode, see also https://github.com/NixOS/nixpkgs/issues/266356#issuecomment-1846417305

**Comment by Ericson2314** [comment]
Clang shell is now broken because a change in how `fileset` works 

> error: access to absolute path '/nix/store' is forbidden in pure eval mode (use '--impure' to override)

**Comment by infinisil** [comment]
@Ericson2314 As above, that only happens with Nix versions <2.13.5, though no idea why such an old version is used there

**Pull request merged**